File: app/src/main/java/org/jellyfin/androidtv/ui/browsing/ByGenreFragment.java
package org.jellyfin.androidtv.ui.browsing;

import org.jellyfin.androidtv.TvApp;
import org.jellyfin.androidtv.data.querying.StdItemQuery;
import org.jellyfin.androidtv.util.Utils;
import org.jellyfin.apiclient.interaction.ApiClient;
import org.jellyfin.apiclient.interaction.Response;
import org.jellyfin.apiclient.model.dto.BaseItemDto;
import org.jellyfin.apiclient.model.querying.ItemSortBy;
import org.jellyfin.apiclient.model.querying.ItemsByNameQuery;
import org.jellyfin.apiclient.model.querying.ItemsResult;
import org.koin.java.KoinJavaComponent;

public class ByGenreFragment extends BrowseFolderFragment {
    @Override
    protected void setupQueries(final IRowLoader rowLoader) {
        if (Utils.getSafeValue(mFolder.getChildCount(), 0) > 0) {
            //Get all genres for this folder
            ItemsByNameQuery genres = new ItemsByNameQuery();
            genres.setParentId(mFolder.getId().toString());
            genres.setSortBy(new String[]{ItemSortBy.SortName});
            if (includeType != null) genres.setIncludeItemTypes(new String[]{includeType});
            genres.setRecursive(true);
            genres.setUserId(TvApp.getApplication().getCurrentUser().getId());
            KoinJavaComponent.<ApiClient>get(ApiClient.class).GetGenresAsync(genres, new Response<ItemsResult>() {
                @Override
                public void onResponse(ItemsResult response) {
                    for (BaseItemDto genre : response.getItems()) {
                        StdItemQuery genreQuery = new StdItemQuery();
                        genreQuery.setParentId(mFolder.getId().toString());
                        genreQuery.setSortBy(new String[]{ItemSortBy.SortName});
                        if (includeType != null)
                            genreQuery.setIncludeItemTypes(new String[]{includeType});
                        genreQuery.setGenres(new String[]{genre.getName()});
                        genreQuery.setRecursive(true);
                        mRows.add(new BrowseRowDef(genre.getName(), genreQuery, 40));
                    }

                    if (mRows.size() < 2) setHeadersState(HEADERS_DISABLED);

                    rowLoader.loadRows(mRows);
                }
            });
        } else {
            setHeadersState(HEADERS_DISABLED);
        }
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/browsing/ByGenreFragment.kt
package org.jellyfin.androidtv.ui.browsing

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jellyfin.androidtv.data.repository.ItemRepository
import org.jellyfin.sdk.api.client.ApiClient
import org.jellyfin.sdk.api.client.extensions.genresApi
import org.jellyfin.sdk.model.api.BaseItemKind
import org.jellyfin.sdk.model.api.ItemSortBy
import org.jellyfin.sdk.model.api.request.GetItemsRequest
import org.koin.android.ext.android.inject

class ByGenreFragment : BrowseFolderFragment() {
	private val apiClient by inject<ApiClient>()

	override suspend fun setupQueries(rowLoader: RowLoader) {
		val childCount = folder?.childCount ?: 0
		if (childCount <= 0) return

		// Get all genres for this folder
		val genresResponse = withContext(Dispatchers.IO) {
			apiClient.genresApi.getGenres(
				parentId = folder?.id,
				sortBy = setOf(ItemSortBy.SORT_NAME),
			).content
		}

		for (genre in genresResponse.items) {
			val itemsRequest = GetItemsRequest(
				parentId = folder?.id,
				sortBy = setOf(ItemSortBy.SORT_NAME),
				includeItemTypes = includeType?.let(BaseItemKind::fromNameOrNull)?.let(::setOf),
				genres = setOf(genre.name.orEmpty()),
				recursive = true,
				fields = ItemRepository.itemFields,
			)
			rows.add(BrowseRowDef(genre.name, itemsRequest, 40))
		}

		rowLoader.loadRows(rows)
	}
}
